File: go/src/RumbleRacing/core/asset/o3d/chunk.go
package o3d

import (
	"encoding/binary"
	"fmt"
	"math"
	"rumble-reader/asset"
)

// Element Header?
type ELHE_Header struct {
	Raw asset.Chunk

	// Relevant Ghidra RE struct data:
	ChildCount       uint16 // 0x00
	MaybeNumTextures int16  // 0x02

	X float32 // 0x48
	Y float32 // 0x4C
	Z float32 // 0x50
	W float32 // 0x54

	RawZDebug   uint32
	RawZAddress int
}

// Element Texture/Translation?
type ELTL_TextureList struct {
	Raw asset.Chunk
}

// Raw element Data: mostly VIF + some texture/other metadata
type ELDA_Data struct {
	Raw asset.Chunk
}

type ObfChunk struct {
	ELHE *ELHE_Header
	ELTL *ELTL_TextureList
	ELDA *ELDA_Data
}

func parseObfChunks(data []byte) ([]ObfChunk, error) {
	var chunks []ObfChunk
	offset := 0

	currentObfChunk := ObfChunk{}
	chunkIndex := 0

	// fmt.Println(hex.Dump(data))
	for offset < len(data) {
		if offset+8 > len(data) {
			return nil, fmt.Errorf("incomplete chunk header at offset %d", offset)
		}
		var magic [4]byte
		copy(magic[:], data[offset:offset+4])

		size := binary.LittleEndian.Uint32(data[offset+4 : offset+8])
		// fmt.Println("OFFSET: ", offset, "SIZE:", size)

		// if size < 8 {
		// 	return nil, fmt.Errorf("invalid chunk size %d for %q", size, magic)
		// }
		chunkEnd := offset + int(size) + 8
		if chunkEnd > len(data) {
			return nil, fmt.Errorf("chunk %q size %d exceeds remaining data", magic, size)
		}
		payload := data[offset:chunkEnd]

		// If the chunk is HEAD, ignore it
		chunk := asset.Chunk{
			Offset:  offset,
			Magic:   magic,
			Size:    size,
			Payload: payload,
		}

		// Ignore HEAD chunk, skip past it
		if chunk.MagicString() == "HEAD" {
			offset = chunkEnd
			continue
		}

		chunkTypeCheck := chunkIndex % 3

		switch chunkTypeCheck {
		case 0:
			if currentObfChunk.ELHE != nil {
				panic("ELHE NOT NIL")
			}
			if chunk.MagicString() != "ELHE" {
				panic("NOT AN ELHE!")
			}

			elhe, err := parseELHE(chunk)
			if err != nil {
				panic("Something went wrong!")
			}
			currentObfChunk.ELHE = elhe

		case 1:
			if currentObfChunk.ELTL != nil {
				panic("ELTL NOT NIL")
			}
			if chunk.MagicString() != "ELTL" {
				panic("NOT AN ELTL!")
			}

			eltl, err := parseELTL(chunk)
			if err != nil {
				panic("Something went wrong!")
			}
			currentObfChunk.ELTL = eltl

		case 2:
			if currentObfChunk.ELDA != nil {
				panic("ELDA NOT NIL")
			}
			if chunk.MagicString() != "ELDA" {
				panic("NOT AN ELDA!")
			}

			elda, err := parseELDA(chunk)
			if err != nil {
				panic("Something went wrong!")
			}
			currentObfChunk.ELDA = elda
		}

		offset = chunkEnd
		chunkIndex++

		// fmt.Println(chunkIndex)
		if currentObfChunk.ELDA != nil && currentObfChunk.ELHE != nil && currentObfChunk.ELTL != nil {
			chunks = append(chunks, currentObfChunk)
			// fmt.Println("RESET")
			currentObfChunk = ObfChunk{}
		}
	}

	return chunks, nil
}

func parseELHE(chunk asset.Chunk) (*ELHE_Header, error) {
	base := 0x8
	elhe := ELHE_Header{
		Raw:              chunk,
		ChildCount:       binary.LittleEndian.Uint16(chunk.Payload[base : base+2]),
		MaybeNumTextures: int16(binary.LittleEndian.Uint16(chunk.Payload[base+0x2 : base+0x2+2])),
		// Unk2:             binary.LittleEndian.Uint16(chunk.Payload[base+0x6 : base+0x6+2]),
		X: math.Float32frombits(binary.LittleEndian.Uint32(chunk.Payload[base+0x48 : base+0x48+4])),
		Y: math.Float32frombits(binary.LittleEndian.Uint32(chunk.Payload[base+0x4C : base+0x4C+4])),
		Z: math.Float32frombits(binary.LittleEndian.Uint32(chunk.Payload[base+0x50 : base+0x50+4])),
		W: math.Float32frombits(binary.LittleEndian.Uint32(chunk.Payload[base+0x54 : base+0x54+4])),

		RawZDebug:   binary.LittleEndian.Uint32(chunk.Payload[base+0x50 : base+0x50+4]),
		RawZAddress: chunk.Offset + base + 0x50,
	}

	return &elhe, nil
}

func parseELTL(chunk asset.Chunk) (*ELTL_TextureList, error) {
	eltl := ELTL_TextureList{
		Raw: chunk,
	}

	return &eltl, nil
}

func parseELDA(chunk asset.Chunk) (*ELDA_Data, error) {
	elda := ELDA_Data{
		Raw: chunk,
	}

	return &elda, nil
}

File: src/RumbleRacing/asset/o3d/chunk.ts
import ArrayBufferSlice from "../../../ArrayBufferSlice";
import { readString } from "../../../util";
import { AssetChunk } from "../chunk";
import { parseVif, VifCommand } from "./vif";

export interface ELHE_Header {
  raw: AssetChunk;
  childCount: number;
  maybeNumTextures: number;
  x: number;
  y: number;
  z: number;
  w: number;
}

export interface ELTL_TextureList {
  raw: AssetChunk;
}

export interface ELDA_Data {
  raw: AssetChunk;
}

export interface ObfChunk {
  elhe: ELHE_Header;
  eltl: ELTL_TextureList;
  elda: ELDA_Data;
}

export function parseELHE(chunk: AssetChunk): ELHE_Header {
  const base = 0x8;
  const p = chunk.payload;
  const view = new DataView(p.buffer, p.byteOffset, p.byteLength);
  return {
    raw: chunk,
    childCount: view.getUint16(base, true),
    maybeNumTextures: view.getInt16(base + 0x2, true),
    x: view.getFloat32(base + 0x48, true),
    y: view.getFloat32(base + 0x4c, true),
    z: view.getFloat32(base + 0x50, true),
    w: view.getFloat32(base + 0x54, true),
  };
}

export function parseELTL(chunk: AssetChunk): ELTL_TextureList {
  return { raw: chunk };
}

export function parseELDA(chunk: AssetChunk): ELDA_Data {
  return { raw: chunk };
}

export function eldaParseVif(elda: ELDA_Data): VifCommand[] {
  return parseVif(elda.raw.payload);
}

export function parseObfChunks(data: Uint8Array): ObfChunk[] {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const chunks: ObfChunk[] = [];
  let offset = 0;

  let currentElhe: ELHE_Header | null = null;
  let currentEltl: ELTL_TextureList | null = null;
  let currentElda: ELDA_Data | null = null;
  let chunkIndex = 0;

  while (offset < data.length) {
    if (offset + 8 > data.length) {
      throw new Error(`incomplete chunk header at offset ${offset}`);
    }

    // OBF fourcc bytes are not reversed like all of the other ones..
    const magicBytes = data.slice(offset, offset + 4);
    const magic = readString(
      ArrayBufferSlice.fromView(magicBytes),
      0,
      4,
      false,
    );
    const size = view.getUint32(offset + 4, true);
    const chunkEnd = offset + size + 8;
    if (chunkEnd > data.length) {
      throw new Error(
        `chunk size ${size} exceeds remaining data at offset ${offset}`,
      );
    }
    const payload = data.slice(offset, chunkEnd);

    const assetChunk: AssetChunk = {
      offset,
      magic: magicBytes,
      size,
      payload,
    };

    if (magic === "HEAD") {
      offset = chunkEnd;
      continue;
    }

    const typeCheck = chunkIndex % 3;
    switch (typeCheck) {
      case 0:
        if (magic !== "ELHE") throw new Error("NOT AN ELHE! Got: " + magic);
        currentElhe = parseELHE(assetChunk);
        break;
      case 1:
        if (magic !== "ELTL") throw new Error("NOT AN ELTL! Got: " + magic);
        currentEltl = parseELTL(assetChunk);
        break;
      case 2:
        if (magic !== "ELDA") throw new Error("NOT AN ELDA! Got: " + magic);
        currentElda = parseELDA(assetChunk);
        break;
    }

    offset = chunkEnd;
    chunkIndex++;

    if (currentElhe && currentEltl && currentElda) {
      chunks.push({ elhe: currentElhe, eltl: currentEltl, elda: currentElda });
      currentElhe = null;
      currentEltl = null;
      currentElda = null;
    }
  }

  return chunks;
}
